### ISSUE TITLE
Escape chars (`\n`) in string literals breaks tests

### ISSUE THREAD

**Issue by Crazybus** [Python-Markdown/markdown#922]
I ran into an odd rendering issue where some links were being rendered inside of a fenced code block.

I managed to narrow it down to situations where there is a `[markdown](url)` followed by an empty line, or a line starting with a `#`. There might be more examples that break the rendering but here some samples:

    ```
    [this works](in-code-block)
    ```

    ```
    [this is fine](in-code-block)
    test
    ```

    ```
    [space after is not ok](in-code-block)
    
    ```

    ```
    [comment after is not ok](in-code-block)
    # not ok
    ```

Example test case that fails:
```
def test_fenced_code_with_markdown_links_space(self):
    self.assertMarkdownRenders(
        self.dedent(
            """
            ```
            [fake-link](http://fake)\n#
            ```
            """
        ),
        self.dedent(
            """
            <pre><code>[fake-link](http://fake)\n#
            </code></pre>
            """
        ),
        extensions=["fenced_code"],
    )
```

Output showing the rendered href.

```
markdown/test_tools.py:64: in assertMarkdownRenders
    self.assertMultiLineEqual(output, expected)
E   AssertionError: '<p>```\n                <a href="http://fak[70 chars]pre>' != '<pre><code>[fake-link](http://fake)\n#\n   [22 chars]pre>'
E   + <pre><code>[fake-link](http://fake)
E   + #
E   +                 </code></pre>- <p>```
E   -                 <a href="http://fake">fake-link</a></p>
E   - <h1></h1>
E   - <pre><code>            ```
E   - </code></pre>
```

P.S. It was actually an unintended markdown formatted link in the code block. In this case it was a gcloud command example where the formatting options happen to also match up with markdown syntax.

```
gcloud compute instances list --filter="zone:${ZONE}" --format="table[no-heading](name)"
```

**Comment by waylan** [comment]
I can't seem to replicate the behavior:

````python
>>> c = '''
... ```
... [space after is not ok](in-code-block)
...
... ```
... '''
>>> markdown.markdown(c, extensions=['fenced_code'])
'<pre><code>[space after is not ok](in-code-block)\n\n</code></pre>'
>>> c = '''
... ```
... [comment after is not ok](in-code-block)
... # not ok
... ```
... '''
>>> markdown.markdown(c, extensions=['fenced_code'])
'<pre><code>[comment after is not ok](in-code-block)\n# not ok\n</code></pre>'
````

So then I tried running your test case and sure enough, I got the same results as you. However I noticed something weird about your test case. Why do you have `\n` in your string literal? Replace the `\n` with an actual newline and the issue goes away:

```python
def test_fenced_code_with_markdown_links_space(self):
    self.assertMarkdownRenders(
        self.dedent(
            """
            ```
            [fake-link](http://fake)
            #
            ```
            """
        ),
        self.dedent(
            """
            <pre><code>[fake-link](http://fake)
            #
            </code></pre>
            """
        ),
        extensions=["fenced_code"],
    )
```

In other words, you had created a bad test. Don't use escape characters in string literals. Use the actual characters. However, for normal strings, the escape characters are fine:

```python
>>> markdown.markdown('```\n[fake-link](http://fake)\n#\n```\n', extensions=['fenced_code'])
'<pre><code>[fake-link](http://fake)\n#\n</code></pre>'
```

Or in a test:

```python
def test_fenced_code_with_markdown_links_space(self):
        self.assertMarkdownRenders(
            '```\n[fake-link](http://fake)\n#\n```\n',
            '<pre><code>[fake-link](http://fake)\n#\n</code></pre>',
            extensions=["fenced_code"],
        )
```

Note that `decent` and string literals are not used there.

**Comment by waylan** [comment]
Why is the use of escape characters breaking the test. Because it is confusing `dedent`. Given the following:

```python
decent(
    '''
    foo\nbar
    '''
)
```

The actual string passed to dedent is:

```
'''
    foo
bar
    '''
```

Notice the inconsistent indentation. Therefore, decent doesn't properly remove that indentation and you get unexpected results. You could do this:

```python
decent(
    '''
    foo\n    bar
    '''
)
```

Note the extra spaces after the `\n`. However, the more levels of indent you get, the more difficult that is to manage properly. Its easier to just use actual carriage returns. Like this:

```python
decent(
    '''
    foo
    bar
    '''
)
```

**Comment by waylan** [comment]
For the record, this is not a bug, but user error.

**Comment by Crazybus** [comment]
Just to follow up on this. This was user error but not for the reason that you thought.

I did not have `extensions=['fenced_code']` set when I was parsing the markdown file. Without this extension enabled I was running into this issue.

When I wasn't able to reproduce it in the unit tests I made a wrong assumption that the `self.dedent` might be causing issues since this seemed like a very subtle issue depending on certain whitespace characters. Adding in the `\n` gave me the same results that I was getting when running this against real files (again without the fenced_code extension enabled).

**Comment by waylan** [comment]
> I did not have `extensions=['fenced_code']` set when I was parsing the markdown file. 

That was my first assumption, but then you provided no evidence to support that. On the contrary, your test case did include it. In any event, I'm glad you worked out the issue.
